### PR TITLE
Store start/length in nodes instead of start/end.  

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**
This speeds up incremental parsing by about 25% as the incremental parser pass only has to update one property on most nodes instead of two.  i.e. now only the start needs to be adjusted (as the length stays the same in any reused node).  Previously both the start and end needed to be adjusted, causing at least two writes on all affected nodes.

This also unifies us with VS and Roslyn's views on what a Span is. The naming is consistent, and the model and operations are the same.

### CODE REVIEW

File: src/compiler/scanner.ts [hunk 1]
```diff
@@ -517,12 +517,12 @@ module ts {
         }
     }
 
-    export function getLeadingCommentRanges(text: string, pos: number): CommentRange[] {
-        return getCommentRanges(text, pos, /*trailing*/ false);
+    export function getLeadingCommentRanges(text: string, pos: number): CommentSpan[] {
```
Comment (DanielRosenwasser) [inline src/compiler/scanner.ts]: Should these be changed to `getLeadingCommentSpans`?

File: src/compiler/utilities.ts [hunk 2]
```diff
@@ -1030,6 +1028,21 @@ module ts {
...
+        return createTextChangeRange(createSpanFromBounds(oldStartN, oldEndN), /*newLength: */newEndN - oldStartN);
+    }
+
+    export function nodeArrayEnd(array: NodeArray<Node>) {
+        var end = array.length === 0 ? array.start : spanEnd(lastOrUndefined(array));
+        if (!array.hasTrailingComma) {
+            return end;
+        }
+
+        Debug.assert(array.length > 0);
```
Comment (DanielRosenwasser) [inline src/compiler/utilities.ts]: Wait, so if there's an empty array with a trailing comma, then...?

Comment (CyrusNajmabadi) [inline src/compiler/utilities.ts]: You can't have an empty array with a trailing comma.  The delimited parsing code won't ever make that.  You only try parsing a comma once you've parsed something real.

File: src/compiler/scanner.ts (hunk 1 above)
Comment (CyrusNajmabadi) [inline src/compiler/scanner.ts]: Sure, i can fix.

File: src/compiler/utilities.ts (hunk 2 above)
Comment (DanielRosenwasser) [inline src/compiler/utilities.ts]: Right; perhaps leave a comment about omitted elements as a reminder.

**Comment by mhegazy** [comment]
let's talk about this first. i think this touches other API changes that we need to agree upon.

**Comment by ahejlsberg** [comment]
I'm not sold on this change. It's pretty invasive and the last thing we need is more churn.

**Comment by mhegazy** [comment]
looks like this has been sitting here for a while. closing it for now.

**Comment by CyrusNajmabadi** [comment]
Reopening.  I still want this to go through.  It just didn't meet the bar for 1.5.  It's definitely a help to incremental parsing.  And i've been consistenly ensuring that the PR is up to date.

**Comment by mhegazy** [comment]
I do not think the value justifies the breaking change at this point.

**Comment by CyrusNajmabadi** [comment]
I disagree.  We have gone through no actual process for providing a supported API for third parties to consume.  At best we've had a "feel free to try, but do not expect any sort of compat guarantee" approach.

I was not under the impression that we would not be breaking things.  If that were the case then i would have actually fought to make this change for 1.5 as i feel it is the right thing to do, both for perf and for VS consistency.  I didn't fight for it then because it didn't feel important enough, and the argument back then was that we could always change it later.

**Comment by basarat** [comment]
I would like to see this (just because its technically a better abstraction). Removing the `end` field should light up any errors for *typed* consumers of the language service API.

**Comment by derekcicerone** [comment]
Fwiw, also cool with the break if it means doing the right thing for the API long-term.

**Comment by DanielRosenwasser** [comment]
:disappointed:

**Pull request closed without merging**